boards: Add support for the Serpente Board

elexx opened the pull request:
The Serpente Boards are small low-cost Atmel SAMD21 based development boards by @arturo182

Signed-off-by: Alexander Falb <fal3xx@gmail.com>

ioannisg reviewed (approved):
overall looks fine to me

nzmichaelh commented on boards/arm/serpente/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+#zephyr_library_sources(board_init.c)
```
Drop the commented-out line

elexx replied:
Good catch. Not sure how I missed that when committing. The line is actually needed to properly setup the onboard led. Otherwise it would be glowing blue as a leftover from SAM-BA bootloader.

nzmichaelh commented on boards/arm/serpente/board_init.c:
```diff
@@ -0,0 +1,37 @@
+/*
```
I see that board_init.c was commented out of the build file, but is included here?

elexx replied:
I will drop this file

nzmichaelh commented on boards/arm/serpente/board_init.c:
```diff
@@ -0,0 +1,37 @@
+	led = device_get_binding(DT_GPIO_LABEL(BLUE_LED, gpios));
+	gpio_pin_configure(led, DT_GPIO_PIN(BLUE_LED, gpios),
+		GPIO_OUTPUT_INACTIVE | DT_GPIO_FLAGS(BLUE_LED, gpios));
+
```
Drop the extra blank link.

nzmichaelh commented on boards/arm/serpente/doc/index.rst:
```diff
@@ -0,0 +1,126 @@
+
+- ATSAMD21E18A ARM Cortex-M0+ processor at 48 MHz
+- 256 KiB flash memory and 32 KiB of RAM
+- extra 4MB SPI flash memory
```
nit: 'Extra' to match the capitalization of the rest of the list.

nzmichaelh commented on boards/arm/serpente/doc/index.rst:
```diff
@@ -0,0 +1,126 @@
+********
+
+The Serpente is a very small low-cost development and prototyping
+board equipped with 4MB flash storage, a PWM enabled RGB led and 6 I/O pins.
```
nit: 4MiB for consistency.

nzmichaelh replied:
Note that Zephyr now has a PWM driver for the SAMD21.  Consider enabling it to drive the LED.

nzmichaelh commented on boards/arm/serpente/pinmux.c:
```diff
@@ -0,0 +1,30 @@
+	ARG_UNUSED(dev);
+
+#ifdef CONFIG_USB_DC_SAM0
+
```
Drop the extra blank line.

nzmichaelh commented on boards/arm/serpente/serpente.dts:
```diff
@@ -0,0 +1,83 @@
+/ {
+	model = "Serpente";
+	compatible =
+		"arturo182,serpente", "atmel,samd21e18a", "atmel,samd21";
```
Use just `compatible = "arturo182,serpente"'`

nzmichaelh commented on boards/arm/serpente/serpente.dts:
```diff
@@ -0,0 +1,83 @@
+
+		boot_partition: partition@0 {
+			label = "sam-ba";
+			reg = <0x00000000 0x2000>;
```
optional: you can now use DT_SIZE_K for the size argument.

nzmichaelh commented on boards/arm/serpente/serpente.dts:
```diff
@@ -0,0 +1,83 @@
+
+		code_partition: partition@2000 {
+			label = "code";
+			reg = <0x2000 0x40000>;
```
The size should be 256 K - the 8 K of the bootloader.

nzmichaelh commented on boards/arm/serpente/serpente.yaml:
```diff
@@ -0,0 +1,10 @@
+arch: arm
+ram: 32
+flash: 256
+toolchain:
```
Consider adding a 'supported' section to get more build coverage.  See the `seeeduino_xiao` for an example.

nzmichaelh commented on boards/arm/serpente/serpente_defconfig:
```diff
@@ -0,0 +1,20 @@
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_UART_SAM0=y
```
The SAM0 drivers are now automatically enabled when the corresponding feature is set, so you can keep the `CONFIG_SERIAL=y` etc and drop the `CONFIG_UART_SAM0=y` etc.

nzmichaelh reviewed (dismissed):
Nice!  Also check out the new seeeduino_xiao config for a more recent SAMD21 port.

elexx commented:
Thanks for the pointers to seeeduino_xiao, I'll have a look.
Do you want me to amend to my original commit or just add a 2nd commit with the requested changes?

nzmichaelh commented:
> Thanks for the pointers to seeeduino_xiao, I'll have a look.
> Do you want me to amend to my original commit or just add a 2nd commit with the requested changes?

Best practice is to amend / fixup and then do a force push.  Make sure you resolve the comments first though, else Github seems to lose them.

elexx commented:
I'm a bit lost regarding PWM.
Because the user LED is ACTIVE_LOW I'd like to set the PWM to PWM_POLARITY_INVERTED (is this the right approach?). However this does not seem to be supported by the driver or I'm doing it wrong:

```
pwmleds {
	compatible = "pwm-leds";
	red_led_pwm: pwm_led_0 {
		pwms = <&tcc0 0 PWM_POLARITY_INVERTED>;
	};
}
```
Results in errors during the build `$ west build samples/basic/blinky_pwm --board=serpente -p`:
```
[...snip...]
serpente.dts.pre.tmp:336.4-30: Warning (pwms_property): /pwmleds/pwm_led_0:pwms: cell 2 is not a phandle reference
serpente.dts.pre.tmp:335.26-337.5: Warning (pwms_property): /pwmleds/pwm_led_0: Missing property '#pwm-cells' in node /soc/interrupt-controller@e000e100 or bad phandle (referred from pwms[2])
devicetree error: <Node /soc/interrupt-controller@e000e100 in 'serpente.dts.pre.tmp'> lacks #pwm-cells
[...snip...]
```

elexx commented:
PWM is now working, however the output is inverted (the LED on the Serpente is ACTIVE_HIGH which is currently not supported by the DTS for SAM0-TCC-PWM). I'm working on another PR to add this feature.

nzmichaelh commented:
> PWM is now working, however the output is inverted (the LED on the Serpente is ACTIVE_HIGH which is currently not supported by the DTS for SAM0-TCC-PWM). I'm working on another PR to add this feature.

Sounds good, I was about to do the same.  Let me know if you run out of time and I can take over the PWM change.

nzmichaelh commented on boards/arm/serpente/serpente_defconfig:
```diff
@@ -0,0 +1,15 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_USB=y
+CONFIG_USE_DT_CODE_PARTITION=y
```
FYI, better support for programming has recently landed.  See https://github.com/zephyrproject-rtos/zephyr/pull/27061 for how I'm enabling it on other SAMD21 boards.

There's no rush though as it needs #27153 to land and a new SDK release for 'west flash' to just work.

elexx replied:
I'll subscribe to #27153 and either change the config lines in this PR or in a follow-up PR, depending on how fast 27153 gets merged.

(I'm already in love with the upcoming changes. No more double-tapping the tiny SMD switch to reboot into bootloader!)

elexx commented:
The Serpente board has a very limited pin count, but the pins are well chosen. It is possible to have 2UART or 1UART+1SPI or 1I2C+1SPI or 1UART+1I2C and so on ([see notes in schematics](https://serpente.solder.party/r2/docs/downloads/)).

What is the best practice in Zephyr for the default sercom definition in the board DTS? The options I was thinking about are:
* Having both available sercoms with status okay, but nothing else in the DTS + overlay for all tests. (Probably the "cleanest" and most time consuming solution I could think of)
* Having some defaults like 1SPI+1I2C in the DTS + overlays for the UART tests (I'm not sure which pinout should be the default)
* Remove the supported flags from yaml (This feels like a quick and dirty hack)

Are there any better ways I missed? Which one do you prefer?

elexx commented:
I just updated the PR and added a few overlay files to configure the board for the tests/samples. However I'm still not sure this is the way to go - please advice!

jukkar commented on samples/net/sockets/http_client/boards/serpente.overlay:
```diff
@@ -0,0 +1,12 @@
+/*
```
How do you use networking and http_client with this board?

elexx replied:
You are right, this overlay file should not be needed. I just deleted it locally and ran the test suite `./scripts/sanitycheck -p serpente`. This also runs "samples/net/sockets/http_client/sample.net.sockets.http_client" which fails with `/home/alexander/Documents/Git/private/zephyr/zephyr/drivers/serial/uart_sam0.c:1031:37: error: 'uart_sam0_driver_api' defined but not used [-Werror=unused-const-variable=]` because no sercom is configured for uart by default.
Why does this sample even get compiled for the serpente board? The serpente.yml does not specify any of the tags from the sample, if I understand correctly. What did I miss?

jukkar replied:
> Why does this sample even get compiled for the serpente board?

Yep, that is the issue here. @nashif any idea?

elexx replied:
Looks like this is a misunderstanding on my side. `sanitycheck` intersects the platform.supported and the testcase.depends_on sets to calculate which testcases should run for which platform. However, the `http_client/sample.yaml` does not have any depends_on definition (this is probably the actual bug here? And maybe the reason for some of the other tests to fail as well.)

elexx replied:
Do you want me to add a `depends_on: netif` to the `http_client/sample.yaml` in this PR or as a separate PR? Or would some other change be better at all?

jukkar replied:
I wonder why the missing `depends_on: netif` is not causing issues with other boards that do not have networking support. Also not all networking samples have dependency mentioned in the `sample.yaml` file. But if that depends_on fixes this issue, you can certainly add a commit to this PR.

elexx replied:
I've been wondering the same, but my understanding of the whole build process is not that good yet. I'd be interested in an answer too.
Alright, I'll amend the PR and add the depends_on.

vanwinkeljan reviewed (dismissed):
Please move `serpente.overlay` files in samples and tests to a separate commit as it has nothing to do with the commit `boards: Add support for the Serpente Board`

elexx commented:
> Please move `serpente.overlay` files in samples and tests to a separate commit as it has nothing to do with the commit `boards: Add support for the Serpente Board`

The tests and samples won't compile without warning (buildkite will fail), if I remove the overlay files. I would also have to remove the `supported` field from `serpente.yaml`, to simply not compile and tests and samples. But then the serpente.yml would not be aligned with the docs that also state what is supported by the board/mcu. I would like to keep this in one PR, but what do you think?

vanwinkeljan commented:
> The tests and samples won't compile without warning (buildkite will fail), if I remove the overlay files. I would also have to remove the `supported` field from `serpente.yaml`, to simply not compile and tests and samples. But then the serpente.yml would not be aligned with the docs that also state what is supported by the board/mcu. I would like to keep this in one PR, but what do you think?

It was just a request to move the overlays to a separate commit, not a separate PR (so one commit to introduce the board and then one commit to add the overlays)

elexx committed:
boards: Add support for the Serpente Board

The Serpente Boards are small low-cost Atmel SAMD21 based
development boards by @arturo182

Signed-off-by: Alexander Falb <fal3xx@gmail.com>

elexx committed:
tests/samples: add samples for Serpente board

Add the supported features section to  serpente.yaml and overlay files
to the corresponding tests and samples.

Signed-off-by: Alexander Falb <fal3xx@gmail.com>

elexx commented:
> It was just a request to move the overlays to a separate commit, not a separate PR (so one commit to introduce the board and then one commit to add the overlays)

Sorry, I misunderstood. I just splittet the commit into two.

elexx commented:
How do you want me to proceed with the failing build_all tests? The serpente only supports 2 different interfaces out of [UART, SPI, I2C] simultaneously, but some tests use all 3 of them. What is the best practice in this scenario?

vanwinkeljan commented:
> How do you want me to proceed with the failing build_all tests? The serpente only supports 2 different interfaces out of [UART, SPI, I2C] simultaneously, but some tests use all 3 of them. What is the best practice in this scenario?

You could blacklist the board for the test that require the 3 interfaces. Have a look at `tests/drivers/build_all/testcase.yaml`, some of the test cases in there already blacklist (`platform_exclude:`) the `frdm_kw41z` board

elexx commented:
I guess I messed up the review resolving. Are there still any changes pending that I missed?

MaureenHelm commented on samples/display/grove_display/boards/serpente.overlay:
```diff
@@ -0,0 +1,12 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&sercom2 {
+	compatible = "atmel,sam0-i2c";
+	clock-frequency = <I2C_BITRATE_FAST>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+};
```
Why aren't all these overlay settings in the main board dts? I know the sercom can be configured as i2c or uart, but once you design a board will you really switch between them?

elexx replied:
I [was debating this](https://github.com/zephyrproject-rtos/zephyr/pull/26960#issuecomment-664792159) myself a few days ago. In the end, I choose to follow the [official pinout diagram](https://serpente.solder.party/r2/docs/pinout/) and [GPIO configuration](https://serpente.solder.party/r2/docs/gpios/) of the Serpente.  
My justification for this was, that the Serpente is a quite small development/prototyping board without a specific use-case in mind. It has just a few pins which provide maximum flexibility - these exact pins are specifically choose, because they provide the most diverse config options (from PWM, touch sensitive inputs, i2c, spi to uart and so on - with only 6 pins!). I didn't want to force a default configuration on the user, which might be unexpected since the official documentation does not describe any default configs.

On the other hand I totally understand that cluttering the code with overlay files is not perfect at all. If you prefer a default configuration of the Serpente in Zephyr, maybe @arturo182 can suggest one?

arturo182 replied:
@elexx put this perfectly, the idea is to not force the user into any specific configuration. TBH, it's even hard to choose which should be the default cause it totally depends on the use-case. If possible, I'd prefer we keep it as flexible as we can.

MaureenHelm replied:
Setting these in the board dts doesn't lock a user into a specific configuration, as you can still override them in an application overlay if desired. I'm not keen on having so many overlays for this board. Can you at least put the console uart in the board dts?

elexx replied:
Thank you for your feedback. I just added a default configuration for the zephyr console and amended the changes to my last commit.

Because I was not able to find this in the documentation or in some other SAMDx based board dts: Is it possible to use the USB connection as zephyr console by default? Without using a sercom at all? I guess this would be the preferred way for the Serpente, because it frees up 2 pins. And imho it also reduces the entry barrier for people new to Zephyr, because only one cable and no additional hardware is required to get started with Zephyr and some quick printk/log debugging.

MaureenHelm replied:
I'm not sure if there are any boards that currently do this, but yes it should be possible. I suggest we go ahead and merge this PR, then you can submit a follow-up PR

elexx commented:
@nzmichaelh I'm sorry, but somewhere on the way to finalize this PR it seems like I forgot to resolve a change request from you. Is there still anything you want me to fix in this PR?

Aside from that, is there anything else I can do to help finish up this PR?

MaureenHelm commented:
@nzmichaelh please revisit

elexx commented:
@nzmichaelh just a friendly reminder that this PR still needs your attention :-)

MaureenHelm commented on boards/arm/serpente/pinmux.c:
```diff
@@ -0,0 +1,66 @@
+#include <drivers/pinmux.h>
+#include <soc.h>
+
+static int board_pinmux_init(struct device *dev)
```
Could you please rebase to the latest master and make the device struct constant? See #24873

elexx replied:
Sure. Done
